velox/functions/prestosql/Zip.cpp

```diff
@@ -0,0 +1,241 @@
+
+  for (auto& arg : inputArgs) {
+    VELOX_USER_CHECK_EQ(
+        arg->type()->kind(),
+        TypeKind::ARRAY,
+        "zip requires arguments of type ARRAY");
+  }
+}
+
+class ZipArrayFunction : public exec::VectorFunction {
+  const static auto MIN_ARITY = 2;
+  const static auto MAX_ARITY = 7;
+
+ public:
+  /// This class implements the zip function.
+  ///
+  /// DEFINITION:
+  /// zip is defined as follows :
+  /// zip(ARRAY[T], ARRAY[U]) -> ARRAY(ROW[T,U])
+  /// where we create a ROW[Ti, Ui] for every ith element in ARRAY[T], ARRAY[U].
+  /// The smaller array is padded with Nulls.
+  ///
+  /// IMPLEMENTATION:
+  ///  1. The general idea is to create a new dictionary vector for each input
+  ///  array vector and enumerating their indexes to create a 1:1 mapping.
+  ///  2. To do this for each row we will determine which is the largest Array
+  ///  and subsequently pad the smaller arrays with Nulls to make them all
+  ///  the same size.
+  ///  3. Then we take the resultant padded vectors together and create one ROW
+  ///  Vector.
+  ///  4. This forms the base to create the final output Array vector, whose
+  ///  Arrays are the size of the largest input Array.
+  ///
+  ///  Note:
+  ///   - This function relies on the inputs being immutable and not being
+  ///   reused, as we reuse the
+  ///     base input elements in the output.
+  ///   - We make no copy's of any constituent elements and are agnostic to
+  ///   types.
+  ///   - For compatibility with Presto a maximum arity is supported.
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    validateType(args);
+    const vector_size_t numInputArrays = args.size();
+
+    std::vector<DecodedVector*> decodedVectors(args.size());
+    std::vector<const ArrayVector*> baseVectors(args.size());
+    std::vector<const vector_size_t*> rawSizes(args.size());
+    std::vector<const vector_size_t*> rawOffsets(args.size());
+    std::vector<const vector_size_t*> indices(args.size());
+
+    for (int i = 0; i < numInputArrays; i++) {
+      exec::LocalDecodedVector holder(context, *args[i].get(), rows);
+      decodedVectors[i] = holder.get();
+      baseVectors[i] = decodedVectors[i]->base()->as<ArrayVector>();
+      rawSizes[i] = baseVectors[i]->rawSizes();
+      rawOffsets[i] = baseVectors[i]->rawOffsets();
+      indices[i] = decodedVectors[i]->indices();
+    }
+
+    // Size of elements in result vector.
+    vector_size_t resultElementsSize = 0;
+
+    // Determine what the size of the resultant elements will be so we can
+    // reserve enough space.
+    auto getMaxArraySize =
+        [&](vector_size_t row) -> std::optional<vector_size_t> {
+      vector_size_t maxSize = 0;
+      for (int i = 0; i < numInputArrays; i++) {
+        if (decodedVectors[i]->isNullAt(row)) {
```

Comment:
Is this possible? zip function has default null behavior, hence, will not be evaluated on rows where one of the inputs is null.

Reply:
No you are right, I wrote a test too for this though. Will remove this check.